crates/map/src/finders/dfs.cairo

```diff
@@ -0,0 +1,179 @@
+//! Depth-First Search algorithm implementation for pathfinding.
+
+// Core imports
+use core::dict::{Felt252Dict, Felt252DictTrait};
+
+// Internal imports
+use origami_map::finders::astar::Astar;
+use origami_map::helpers::bitmap::Bitmap;
+use origami_map::helpers::seeder::Seeder;
+use origami_map::types::node::{Node, NodeTrait};
+use origami_map::types::direction::{Direction, DirectionTrait};
+
+/// DFS implementation for pathfinding
+#[generate_trait]
+pub impl DFS of DFSTrait {
+    /// Searches for a path from 'from' to 'to' on the given grid using DFS
+    ///
+    /// # Arguments
+    /// * `grid` - The grid represented as a felt252
+    /// * `width` - The width of the grid
+    /// * `height` - The height of the grid
+    /// * `from` - The starting position
+    /// * `to` - The target position
+    ///
+    /// # Returns
+    /// A Span<u8> representing the path from 'from' to 'to', or an empty span if no path exists
+    #[inline]
+    fn search(grid: felt252, width: u8, height: u8, from: u8, to: u8) -> Span<u8> {
+        // [Check] The start and target are walkable
+        if Bitmap::get(grid, from) == 0 || Bitmap::get(grid, to) == 0 {
+            return array![].span();
+        }
+
+        // [Effect] Initialize the start and target nodes
+        let mut start = NodeTrait::new(from, 0, 0, 0);
+        let target = NodeTrait::new(to, 0, 0, 0);
+
+        // [Effect] Initialize the stack and the visited nodes
+        let mut stack: Array<Node> = array![start];
+        let mut visited: Felt252Dict<bool> = Default::default();
+        let mut parents: Felt252Dict<u8> = Default::default();
+
+        // [Compute] DFS until the target is reached or stack is empty
+        let mut path_found = false;
+        while let Option::Some(current) = stack.pop_front() {
+            // [Check] Stop if we reached the target
+            if current.position == target.position {
+                path_found = true;
+                break;
+            }
+
+            // [Check] Skip if already visited
+            if visited.get(current.position.into()) {
+                continue;
+            }
+
+            // [Effect] Mark as visited
+            visited.insert(current.position.into(), true);
+
+            // [Compute] Evaluate the neighbors for all 4 directions
+            let seed = Seeder::shuffle(grid, current.position.into());
+            let mut directions = DirectionTrait::compute_shuffled_directions(seed);
+            while directions != 0 {
+                let direction = DirectionTrait::pop_front(ref directions);
+                if Astar::check(grid, width, height, current.position, direction, ref visited) {
+                    let neighbor_position = direction.next(current.position, width);
+                    parents.insert(neighbor_position.into(), current.position);
+                    let neighbor = NodeTrait::new(neighbor_position, current.position, 0, 0);
+                    stack.append(neighbor);
```

Comment:
What you have done is kind of a random search, searching circularly neighbour after neighbour.
You should probably forget the while loops and try a recursive approach to have a DFS algo.